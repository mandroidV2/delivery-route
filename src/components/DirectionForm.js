import React, { Component } from 'react';
import * as ApiManager from '../api/ApiManager'

/** Component is used to create the direction input form
 *  @author Manish Agrawal */
class DirectionForm extends Component {

    constructor(props) {
        super(props)
        this.state = {
            source : '',
            destination: '',
            totalDistance : 0,
            totalTime : 0,
            reset: true
        }
        this.handleLocationChange = this.handleLocationChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleResetClick = this.handleResetClick.bind(this);
    }

    /**
     * Method is used to change the state of location inputs
     * @param {event on element} event 
     */
    handleLocationChange(event) {
        const value = event.target.value;
        const name = event.target.name;
        this.setState({
            [name]: value
        });
    }

    /**
     * Method is used to handle the submit event
     * @param {submit event} event 
     */
    handleSubmit(event) {
        event.preventDefault();
        // call the api to find the paths
        ApiManager.submitLocation({}, (status, response, error) => {
            console.log('result : ' , status);console.log('response : ' , response);console.log('error : ' , error);
            if (error) {
                this.props.onApiErrorOccured();
                this.setState({ reset: false });
            } else {
                ApiManager.getDrivingRoute(response.token, (status, response, error) => {
                    this.setState({ reset: false });
                    console.log('result : ' , status);console.log('response : ' , response);console.log('error : ' , error);
                    if(error) {
                        this.props.onApiErrorOccured();
                    } else {
                        if (response) {
                            switch(response.status) {
                                case 'in progress':
                                    this.props.onProgressStateFound(response.status);
                                    break;
                                case 'failure':
                                    this.props.onRouteFailure(response.error);
                                    break;
                                case 'success':
                                    this.setState({
                                        totalDistance : response.total_distance,
                                        totalTime: response.total_time
                                    });
                                    this.props.onDrivingRouteFound(response);
                                    break;
                                default:
                                    break;
                            }
                          }
                    }
                })
            }
        });
    }

    /**
     * Method is used to handle reset button click
     */
    handleResetClick() {
        this.setState({
            source : '',
            destination: '',
            totalDistance: 0,
            totalTime: 0,
            reset: true
        });
        this.props.onFormReset();
    }

    /**
     * Method is used to clear the location input
     * @param {input state name} name 
     */
    clearInput(name) {
        this.setState({
            [name]: ''
        });
    }

    render() {
        return (
            <div>
                <form onSubmit={this.handleSubmit}>
                    <label>Starting Location</label>
                    &nbsp;&nbsp;
                    <input type = "text" 
                        name = "source"
                        value = {this.state.source} 
                        onChange={this.handleLocationChange}  required />
                    
                    {this.state.source ? 
                        <img className="cross-img"
                            onClick={() => this.clearInput('source')}
                            alt="Clear starting location"
                            src="https://upload.wikimedia.org/wikipedia/commons/3/36/Close%2C_Web_Fundamentals.svg" />
                    :  ''}

                    <br />
                    <br />
                    
                    <label>Drop-off Location</label>
                    &nbsp;&nbsp;
                    <input type = "text"
                        name = "destination"
                        value = {this.state.destination}
                        onChange={this.handleLocationChange} required />
                   
                    {this.state.destination ? 
                        <img className="cross-img" 
                            onClick={() => this.clearInput('destination')}
                            alt="Clear drop-off location"
                            src="https://upload.wikimedia.org/wikipedia/commons/3/36/Close%2C_Web_Fundamentals.svg" />
                    :  ''}

                    <br />
                    <br />

                    {this.state.totalDistance !== 0 ?
                        <div>
                            Total distance : <span>{this.state.totalDistance}</span>
                            <br/>
                            Total time : <span>{this.state.totalTime}</span>
                        </div> : ''}


                    <br />

                    <input type = "submit" value={this.state.reset ? 'Submit' : 'Re-Submit'} />
                    &nbsp;&nbsp;
                    <input type = "reset" onClick={this.handleResetClick} />
                </form>
            </div>
        );
    }
}

export default DirectionForm