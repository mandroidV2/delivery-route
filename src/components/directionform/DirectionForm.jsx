import React, { Component } from "react";
import * as ApiManager from "../../api/ApiManager";
import Loader from "../loader/Loader.jsx";
import maps from "../map/mapConfig";

/** Component is used to create the direction input form
 *  @author Manish Agrawal */
class DirectionForm extends Component {
  constructor() {
    super();
    this.state = {
      source: "",
      destination: "",
      totalDistance: 0,
      totalTime: 0,
      reset: true,
      isLoading: false
    };
    this.handleLocationChange = this.handleLocationChange.bind(this);
    this.autoComplete = this.autoComplete.bind(this);
  }

  componentDidMount() {
    this.autoComplete();
  }

  /**
   * Method is used to implement autocomplete functionality
   */
  autoComplete = async () => {
    const maps = await this.props.maps();
    const _this = this;
    new maps.places.Autocomplete(this.refs.sourceInput).addListener(
      "place_changed",
      function() {
        _this.setState({ source: this.getPlace().formatted_address });
      }
    );
    new maps.places.Autocomplete(this.refs.destinationInput).addListener(
      "place_changed",
      function() {
        _this.setState({ destination: this.getPlace().formatted_address });
      }
    );
  };

  /**
   * Method is used to change the state of location inputs
   * @param {event on element} event
   */
  handleLocationChange(event) {
    const value = event.target.value;
    const name = event.target.name;
    this.setState({
      [name]: value
    });
  }

  /**
   * Method is used to handle the submit event
   * @param {submit event} event
   */
  handleSubmit = async event => {
    event.preventDefault();
    this.setState({ isLoading: true });

    // gather form data
    var locationData = {};
    locationData.source = this.state.source;
    locationData.destination = this.state.destination;

    // fire the submit location api
    var result = await ApiManager.submitLocation(locationData).catch(err => {
      this.props.onApiErrorOccured();
    });
    if (result && result.data && result.data.token) {
      var direction = await ApiManager.getDrivingRoute(result.data.token).catch(
        err => this.props.onApiErrorOccured()
      );
      if (direction && direction.data) {
        let response = direction.data;
        switch (response.status) {
          case "in progress":
            this.props.onProgressStateFound(response.status);
            break;
          case "failure":
            this.props.onRouteFailure(response.error);
            break;
          case "success":
            this.setState({
              totalDistance: response.total_distance,
              totalTime: response.total_time
            });
            this.props.onDrivingRouteFound(response);
            break;
          default:
            break;
        }
      }
    }
    this.setState({ reset: false, isLoading: false });
  };

  /**
   * Method is used to handle reset button click
   */
  handleResetClick() {
    this.setState({
      source: "",
      destination: "",
      totalDistance: 0,
      totalTime: 0,
      reset: true,
      isLoading: false
    });
    this.props.onFormReset();
  }

  /**
   * Method is used to clear the location input
   * @param {input state name} name
   */
  clearInput(name) {
    this.setState({
      [name]: ""
    });
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit.bind()}>
          <label>Starting Location</label>
          <br />
          <input
            type="text"
            name="source"
            ref="sourceInput"
            value={this.state.source}
            onChange={this.handleLocationChange}
            required
          />

          <img
            className="cross_source cross"
            onClick={() => this.clearInput("source")}
            alt="Clear starting location"
            style={{ display: this.state.source ? "inline-block" : "none" }}
            src="https://upload.wikimedia.org/wikipedia/commons/3/36/Close%2C_Web_Fundamentals.svg"
          />

          <br />
          <br />

          <label>Drop-off Location</label>
          <br />
          <input
            type="text"
            name="destination"
            ref="destinationInput"
            value={this.state.destination}
            onChange={this.handleLocationChange}
            required
          />

          <img
            className="cross_destination cross"
            onClick={() => this.clearInput("destination")}
            alt="Clear drop-off location"
            style={{
              display: this.state.destination ? "inline-block" : "none"
            }}
            src="https://upload.wikimedia.org/wikipedia/commons/3/36/Close%2C_Web_Fundamentals.svg"
          />

          <br />
          <br />

          <div
            style={{
              display: this.state.totalDistance !== 0 ? "inline-block" : "none"
            }}
          >
            Total distance : <span>{this.state.totalDistance}</span>
            <br />
            Total time : <span>{this.state.totalTime}</span>
          </div>

          <br />
          <br />

          {this.state.isLoading ? (
            <Loader />
          ) : (
            <div>
              <input
                type="submit"
                value={this.state.reset ? "Submit" : "Re-Submit"}
              />
              &nbsp;&nbsp;
              <input type="reset" onClick={this.handleResetClick.bind(this)} />
            </div>
          )}
        </form>
      </div>
    );
  }
}

DirectionForm.defaultProps = {
  maps
};

export default DirectionForm;
