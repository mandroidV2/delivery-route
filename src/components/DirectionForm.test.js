import React from 'react';
import { shallow } from 'enzyme';
import DirectionForm from './DirectionForm';

/** Test case is used to check the map container component  */
describe('DirectionForm', () => {
    const component = shallow( <DirectionForm /> );
    it('should be defined', () => {
        expect(component).toBeDefined();
    });
    it('should render correctly', () => {
        expect(component).toMatchSnapshot();
    });
});