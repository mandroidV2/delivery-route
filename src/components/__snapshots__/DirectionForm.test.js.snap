// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectionForm should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DirectionForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
      >
        <label>
          Starting Location
        </label>
          
        <input
          name="source"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        
        <br />
        <br />
        <label>
          Drop-off Location
        </label>
          
        <input
          name="destination"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        
        <br />
        <br />
        
        <br />
        <input
          type="submit"
          value="Submit"
        />
          
        <input
          onClick={[Function]}
          type="reset"
        />
      </form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label>
            Starting Location
          </label>,
          "  ",
          <input
            name="source"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />,
          "",
          <br />,
          <br />,
          <label>
            Drop-off Location
          </label>,
          "  ",
          <input
            name="destination"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />,
          "",
          <br />,
          <br />,
          "",
          <br />,
          <input
            type="submit"
            value="Submit"
          />,
          "  ",
          <input
            onClick={[Function]}
            type="reset"
          />,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Starting Location",
          },
          "ref": null,
          "rendered": "Starting Location",
          "type": "label",
        },
        "  ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "source",
            "onChange": [Function],
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Drop-off Location",
          },
          "ref": null,
          "rendered": "Drop-off Location",
          "type": "label",
        },
        "  ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "destination",
            "onChange": [Function],
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "type": "submit",
            "value": "Submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "  ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onClick": [Function],
            "type": "reset",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
        >
          <label>
            Starting Location
          </label>
            
          <input
            name="source"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
          
          <br />
          <br />
          <label>
            Drop-off Location
          </label>
            
          <input
            name="destination"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
          
          <br />
          <br />
          
          <br />
          <input
            type="submit"
            value="Submit"
          />
            
          <input
            onClick={[Function]}
            type="reset"
          />
        </form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Starting Location
            </label>,
            "  ",
            <input
              name="source"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />,
            "",
            <br />,
            <br />,
            <label>
              Drop-off Location
            </label>,
            "  ",
            <input
              name="destination"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />,
            "",
            <br />,
            <br />,
            "",
            <br />,
            <input
              type="submit"
              value="Submit"
            />,
            "  ",
            <input
              onClick={[Function]}
              type="reset"
            />,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Starting Location",
            },
            "ref": null,
            "rendered": "Starting Location",
            "type": "label",
          },
          "  ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "source",
              "onChange": [Function],
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Drop-off Location",
            },
            "ref": null,
            "rendered": "Drop-off Location",
            "type": "label",
          },
          "  ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "destination",
              "onChange": [Function],
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "type": "submit",
              "value": "Submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "  ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onClick": [Function],
              "type": "reset",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
